File: codeJava/src/me/imtt/sorting/advance/merge/mergeSortBu.java
package me.imtt.sorting.advance.merge;

/**
 * 自底向上进行归并排序
 */
public class mergeSortBu {
    /**
     * 辅助数组
     */
    private static Comparable[] aux;

    /**
     * 进行 lgN 次两两归并
     */
    public static void sort(Comparable[] arr) {
        int n = arr.length;
        aux = new Comparable[n];
        for (int sz = 1; sz < n; sz += sz) {
            for (int l = 0; l < n - sz; l += sz + sz) {
                //对于arr[mid] <= arr[mid+1]的情况,不进行merge
                if (arr[l + sz - 1].compareTo(arr[l + sz]) > 0) {
                    merge(arr, l, l + sz - 1, Math.min(l + sz + sz - 1, n - 1));
                }
            }
        }
    }

    /**
     * 将 arr[l...mid] 和 arr[mid+1...r] 两部分进行归并
     */
    private static void merge(Comparable[] arr, int l, int mid, int r) {
        int i = l;
        int j = mid + 1;

        System.arraycopy(arr, l, aux, l, r - l + 1);

        for (int k = l; k <= r; k++) {
            if (i > mid) {
                //左半部分元素已经处理完毕
                arr[k] = aux[j++];
             } 
            else if (j > r) {
                //右半部分元素已经处理完毕
                arr[k] = aux[i++]; 
            } else if (aux[i].compareTo(aux[j]) < 0) {
                arr[k] = aux[i++];
            } else {
                arr[k] = aux[j++];
            }
        }
    }
}

File: codePython/02-Sorting-Advance/merge_sort/merge_sort_bu.py
"""
自底向上进行归并排序（升序）
"""
import copy


def sort(lst):
    """
    进行 lgN 次两两归并
    """
    n = len(lst)
    sz = 1  # sz 子数组大小
    while sz < n:
        l = 0  # l 子数组索引
        while l < n - sz:
            # 对于 lst[mid] <= lst[mid + 1]的情况, 不进行merge
            if lst[l + sz - 1] > lst[l + sz]:
                _merge(lst, l, l + sz - 1, min(l + sz + sz - 1, n - 1))
            l += sz + sz
        sz += sz


def _merge(lst, l, mid, r):
    """
    将 lst[l...mid] 和 lst[mid+1...r] 两部分进行归并
    """
    aux = copy.deepcopy(lst[l:r + 1])
    i, j = l, mid + 1
    for k in range(l, r + 1):
        if i > mid:  # 左半部分元素已经处理完毕
            lst[k] = aux[j - l]
            j += 1
        elif j > r:  # 右半部分元素已经处理完毕
            lst[k] = aux[i - l]
            i += 1
        elif aux[i - l] < aux[j - l]:
            lst[k] = aux[i - l]
            i += 1
        else:
            lst[k] = aux[j - l]
            j += 1

